File: 2021/12/04/main.go
package main

type tree struct {
	next   [26]*tree
	isWord bool
}

type StreamChecker struct {
	tree   *tree
	stream []int
}

func Constructor(words []string) StreamChecker {
	m := 0
	t := &tree{}
	for _, w := range words {
		m = max(m, len(w))
		t.insert(w)
	}
	return StreamChecker{
		tree:   t,
		stream: make([]int, 0, 1024),
	}
}

func (this *StreamChecker) Query(letter byte) bool {
	this.stream = append(this.stream, int(letter-'a'))
	n, t := len(this.stream), this.tree
	for i := n - 1; i >= 0; i-- {
		index := this.stream[i]
		if t.next[index] == nil {
			return false
		}
		t = t.next[index]
		if t.isWord {
			return true
		}
	}
	return false
}

func (t *tree) insert(word string) {
	n := len(word)
	// reversely insert
	for i := n - 1; i >= 0; i-- {
		index := int(word[i] - 'a')
		if t.next[index] == nil {
			t.next[index] = &tree{}
		}
		t = t.next[index]
	}
	t.isWord = true
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

/**
 * Your StreamChecker object will be instantiated and called as such:
 * obj := Constructor(words);
 * param_1 := obj.Query(letter);
 */

File: 2021/12/04/main.py

class StreamChecker:

    def __init__(self, words: List[str]):
        self.tree={}
        for word in words:
            curnode=self.tree
            word=word[::-1]
            for ch in word:
                if ch not in curnode:
                    curnode[ch]={}
                curnode=curnode[ch]
            curnode['#']=1
        self.pre=''
 
    def query(self, letter: str) -> bool:
        self.pre=self.pre+letter
        curnode=self.tree
        for i in range(0,len(self.pre)):
            if self.pre[-i-1] not in curnode:
                break
            curnode=curnode[self.pre[-i-1]]
            if '#' in curnode:
                return True
        return False
        


# Your StreamChecker object will be instantiated and called as such:
# obj = StreamChecker(words)
# param_1 = obj.query(letter)